rosidl_typesupport_opendds_cpp/include/rosidl_typesupport_opendds_cpp/requester.h

```diff
@@ -0,0 +1,66 @@
+#ifndef ROSIDL_TYPESUPPORT_OPENDDS_CPP__REQUESTER_H_
+#define ROSIDL_TYPESUPPORT_OPENDDS_CPP__REQUESTER_H_
+
+#include <string>
+#include "dds/DCPS/Service_Participant.h"
+#include "dds/DCPS/Marked_Default_Qos.h"
+#include "dds/DdsDcpsC.h"
+
+#include "requester_parameters.h"
+
+namespace rosidl_typesupport_opendds_cpp
+{
+
+  template <typename TReq, typename TRep>
+  class Requester {
+  public:
+    typedef TReq RequestType;
+
+    typedef TRep ReplyType;
+
+    typedef typename OpenDDS::DCPS::DDSTraits<TReq>::DataWriterType RequestDataWriter;
+    typedef typename OpenDDS::DCPS::DDSTraits<TReq>::DataReaderType RequestDataReader;
+
```

Comment:
Fixed

Reply:
What I meant was that the template parameters TReq and TRep should only be used to define RequestType and ReplyType, and then RequestType and ReplyType should be used consistently for the rest of the class declaration and implementation. (They can even be used by other objects using this class.) 

The first few lines are fine:

template <typename TReq, typename TRep>
class Requester {
public:
typedef TReq RequestType;
typedef TRep ReplyType;

The following lines should be:
typedef typename OpenDDS::DCPS::DDSTraits\<RequestType\>::DataWriterType RequestDataWriter;
typedef typename OpenDDS::DCPS::DDSTraits\<ReplyType\>::DataReaderType ReplyDataReader;

Reply:
It is different from the OMG Reqester class definition. What specific problem would we face if we accept the code as it is?

Reply:
These two lines appear in the files from the OMG.  I think we should have something equivalent here.
```
    typedef TReq RequestType;
    typedef TRep ReplyType;
```

Reply:
That part is easy - I can uncomment it back. I think our bigger question is if there is a benefit to using RequestType / ReplyType instead of TReq / TRep internally in the class after we have these typedefs. I am not sure I see one...

Reply:
RequestType is not used "internally" in the file `request_reply.h` from OMG

Reply:
Right. My understanding the benefit of this typedef would be limited to the class users, if they ever need to use these types (which should not be the case with RMW: when RMW needs reader / writer, it goes through the typesupport API). I'll let Jiang comment on his suggestion to avoid misinterpretation.

Reply:
1. The difference in keeping these two typedef's or not could be whether we implemented DDS-RPC interface correctly or not. Since they are defined in public section, I'm not yet convinced that they will not be used by other classes. Could you list all the files that include this header file?

2. The difference in using a type definition consistently or not is simply consistency, usability, and maintainability. The relationship between a type parameter and a type is similar to the relationship between a type and a value. After a while, you will feel as strongly to use type definition consistently as to use nullptr consistently instead of using 0 occasionally.